pkg/frontend/querymiddleware/querysharding.go

```diff
@@ -145,15 +146,21 @@ func (s *querySharding) Do(ctx context.Context, r Request) (Response, error) {
 	queryStats := stats.FromContext(ctx)
 	queryStats.AddShardedQueries(uint32(shardingStats.GetShardedQueries()))
 
-	r = r.WithQuery(shardedQuery)
-	shardedQueryable := newShardedQueryable(r, s.next)
+	shardedR := r.WithQuery(shardedQuery)
+	shardedQueryable := newShardedQueryable(shardedR, s.next)
 
-	qry, err := newQuery(ctx, r, s.engine, lazyquery.NewLazyQueryable(shardedQueryable))
+	qry, err := newQuery(ctx, shardedR, s.engine, lazyquery.NewLazyQueryable(shardedQueryable))
 	if err != nil {
 		return nil, apierror.New(apierror.TypeBadData, err.Error())
 	}
 
 	res := qry.Exec(ctx)
+	for _, anno := range res.Warnings {
+		// Rerun this without query sharding in the event the sharded version results in this monotonicity bug.
+		if errors.Is(annotations.ExtractAnnoError(anno), annotations.HistogramQuantileForcedMonotonicityInfo) {
+			return s.next.Do(ctx, r)
```

Comment:
can you also add some span log to the trace so we know this happened?

Reply:
how often is this happening? my understanding is that it's very rare. Can we can log a warning as well when this happens?

Reply:
I'm not sure how often it's happening actually. Once this warning makes it to production we'll have a better idea, and with this fix we can check the logs (added the warn log). If it turns out to be very common, we can consider disabling query sharding for this type of query. What do you mean by span log to the trace?